server/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java

```diff
@@ -134,31 +130,39 @@ void executeSearch(
         final ActionListener<MultiSearchResponse> listener,
         final long relativeStartTime
     ) {
+        /*
+         * The number of times that we poll an item from the queue here is the minimum of the number of requests and the maximum number
+         * of concurrent requests. At first glance, it appears that we should never poll from the queue and not obtain a request given
+         * that we only poll here no more times than the number of requests. However, this is not the only consumer of this queue as
+         * earlier requests that have already completed will poll from the queue too, and they could complete before later polls are
+         * invoked here. Thus, it can be the case that we poll here and the queue was empty.
+         */
         SearchRequestSlot request = requests.poll();
-        if (request == null) {
-            /*
-             * The number of times that we poll an item from the queue here is the minimum of the number of requests and the maximum number
-             * of concurrent requests. At first glance, it appears that we should never poll from the queue and not obtain a request given
-             * that we only poll here no more times than the number of requests. However, this is not the only consumer of this queue as
-             * earlier requests that have already completed will poll from the queue too and they could complete before later polls are
-             * invoked here. Thus, it can be the case that we poll here and the queue was empty.
-             */
-            return;
+        // If we have another request to execute, we execute it. If the execution forked #doExecuteSearch will return false and will
+        // recursively call this method again eventually. If it did not fork and was able to execute the search right away #doExecuteSearch
+        // will return true, in which case we continue and run the next search request here.
+        while (request != null && doExecuteSearch(requests, responses, responseCounter, relativeStartTime, request, listener)) {
```

Comment:
There's a mild change in behaviour here. We will now potentially run slightly longer tasks on the transport threads if we have a bunch of search requests that get a response right away back to back. It seems to me this is a rare scenario and if encountered, then it also is quite cheap on a per-request basis.
=> It seemed fine to me to not worry about this.

Reply:
We know for a fact that Search is sometimes blocking transport threads (especially in can_match & when there are many indices, etc.). We have other issues describing this particular kind of bug. 

Wouldn't it be prudent to fix that FIRST and then adjust this?

Reply:
That is a fair point. You could argue this both ways though and go with: If there's so many non-forking searches that not forking blocks us here, then forking will result in horrible latency from essentially forking N-searches (N-concurrency-level) times relative to this solution.
Forking isn't free at all, especially to `GENERIC`, so there's always going to be a cutoff here where one or the other behaves better (the linked issue shows a O(100ms) difference here).

Lets discuss this on another channel, ping incoming there.

Reply:
I'm starting to feel like this is good to go now that #110860 is merged. The can_match situation is a lot less relevant with that fix alone and also on second thought :D how do I even get into the situation of many indices and this thing resolving on the current thread? The indices targeted would all have to reside on the current node which is not realistic?